File: YanDong/week6/Populating_Next_Right_Pointers_in_Each_Node.c
struct TreeLinkNode {
    TreeLinkNode *left;
    TreeLinkNode *right;
    TreeLinkNode *next;
};
/*一次AC...*/
void connect(struct TreeLinkNode *root) {
    if(root == NULL)
        return root;
    struct TreeLinkNode *leftmost = root;
    while(leftmost->left != NULL){
        struct TreeLinkNode * curNode = leftmost;
        while(curNode != NULL){
            curNode->left->next = curNode->right;
            if(curNode->next != NULL){
                curNode->right->next = curNode->next->left;
            }
            curNode = curNode->next;
        }
        leftmost = leftmost->left;
    }
}
int main(){
    return 0;
}

File: SheYi/2016-01-10-week6/PopulatingNextRightPointersinEachNode.java
package algorithm;

public class PopulatingNextRightPointersinEachNode {
	//I AC 0ms
/*	   public void connect(TreeLinkNode root) {
		   if(root==null) return ;
	          if(root.left!=null){
	        	  root.left.next=root.right;
	        	  if(root.next==null)
	        		  root.right.next=null; 
	        	  else
	        		  root.right.next=root.next.left;
	          }
	          connect(root.left);
	          connect(root.right);
	          
	    }*/
	//II 
    public void connect(TreeLinkNode root) {
       if(root==null) return ;
       if(root.right!=null){
    	   if(root.next==null){
    		   root.right.next=null;
    	   }else{
    		  root.right.next=getNextNode(root);
    	   }
    	   connect(root.right);
       }
       if(root.left!=null){
    	  if(root.right!=null){
    		  root.left.next=root.right;
    	  }else{
    		  root.left.next=getNextNode(root);
    	  }
    	  connect(root.left);
       }
         
    }
    
    public TreeLinkNode getNextNode(TreeLinkNode cur){
    	 TreeLinkNode temp=null;
    	 temp=cur.next;
    	while(temp!=null){
    		if(temp.left!=null) return temp.left;
    		else if(temp.right!=null) return temp.right;
    		else{
    			temp=temp.next;
    		}
    	}
    	return null;
    }
}
